pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,167 +5,174 @@ package watchers
+					}
+					if needUpdate {
+						cache[event.Key] = event.Object.DeepCopy()
+					}
+				case resource.Delete:
+					k.onCiliumNodeDelete(event.Object)
+					delete(cache, event.Key)
+				}
+				event.Done(err)
+			}
+		}()
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
 
-		// Set the ciliumNodeStore as nil so that any attempts of getting the
-		// CiliumNode are performed with a request sent to kube-apiserver
-		// directly instead of relying on an outdated version of the CiliumNode
-		// in this cache.
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = nil
-		k.ciliumNodeStoreMU.Unlock()
+			store, err := k.resources.CiliumNode.Store(subCtx)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					log.WithError(err).Warning("unable to retrieve CiliumNode local store, going to query kube-apiserver directly")
+				}
+				return
+			}
 
-		close(isConnected)
+			k.ciliumNodeStoreMu.Lock()
+			k.ciliumNodeStore = store
+			k.ciliumNodeStoreMu.Unlock()
 
-		k.cancelWaitGroupToSyncResources(apiGroup)
-		k.k8sAPIGroups.RemoveAPI(apiGroup)
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+			<-subCtx.Done()
 
-		log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+			k.ciliumNodeStoreMu.Lock()
+			k.ciliumNodeStore.Release()
+			k.ciliumNodeStore = nil
+			k.ciliumNodeStoreMu.Unlock()
+		}()
+
+		select {
+		case <-kvstore.Connected():
+			log.Info("Connected to key-value store, stopping CiliumNode watcher")
+			cancel()
+			k.cancelWaitGroupToSyncResources(apiGroup)
+			k.k8sAPIGroups.RemoveAPI(apiGroup)
+			wg.Wait()
+		case <-ctx.Done():
+			cancel()
+			wg.Wait()
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-kvstore.Client().Disconnected():
+			log.Info("Disconnected from key-value store, restarting CiliumNode watcher")
+		}
 	}
 }
 
-// GetCiliumNode returns the CiliumNode "nodeName" from the local store. If the
-// local store is not initialized then it will fallback retrieving the node
-// from kube-apiserver.
-func (k *K8sWatcher) GetCiliumNode(ctx context.Context, nodeName string) (*cilium_v2.CiliumNode, error) {
-	var (
-		err                      error
-		nodeInterface            interface{}
-		exists, getFromAPIServer bool
-	)
-	k.ciliumNodeStoreMU.RLock()
-	// k.ciliumNodeStore might not be set in all invocations of GetCiliumNode,
-	// for example, during Cilium initialization GetCiliumNode is called from
-	// WaitForNodeInformation, which happens before ciliumNodeStore,
-	// so we will fallback to perform an API request to kube-apiserver.
-	if k.ciliumNodeStore == nil {
-		getFromAPIServer = true
-	} else {
-		nodeInterface, exists, err = k.ciliumNodeStore.GetByKey(nodeName)
+func (k *K8sWatcher) onCiliumNodeInsert(ciliumNode *cilium_v2.CiliumNode) bool {
+	if k8s.IsLocalCiliumNode(ciliumNode) {
+		return false
+	}
+	n := types.ParseCiliumNode(ciliumNode)
+	k.nodeDiscoverManager.NodeUpdated(n)
```

Comment:
Since this seems to be the only thing happening here wouldn't it be easy enough to make `NodeDiscovery` subscribe to `Resource[*CiliumNode]` and just remove `pkg/k8s/watchers/cilium_node.go` altogether? `GetCiliumNode` is also trivial to implement around `Resource[*CiliumNode].Store()`.

Reply:
> make NodeDiscovery subscribe to Resource[*CiliumNode]

For completeness, the node discovery module only needs to subscribe to Resource[LocalCiliumNode], which doesn't need to be stopped.